### PR TITLE
conformance: deepcopiable doesn't imply distinct ones

### PR DESCRIPTION
**Pull request by fingolfin**
Splitting my PR, so we can better focus on the various issues.

### CODE REVIEW

**Comment by fingolfin** [comment]
PR #41 contains a discussion of aspects of this PR, which one should probably read before continuing to read here.

There @kalmarek suggested to adjust the definition of `_is_deepcopiable ` to something like `_is_deepcopiable(g::GroupElement) = !isbits(g) && all(nf->isdefined(g, nf), fieldnames(typeof(g)))`.


But I'd like to take a step and ask: Why do we even care...
1. whether `deepcopy(g) !== deepcopy(g)` for "deepcopyable" group element types;
2. whether `one(G) !== one(g)` in general?

Why are either of these desirable, and worth jumping through hoops? 

@kalmarek wrote on PR #41:
> The purpose is to make sure that if a group element is deepcopiable then deepcopy_internal (which the user must implement) works as intended = a deepcopy of a deepcopiable element is truly distinct.

First off, I don't see how that justifies the `one(G) !== one(g)` check: no `deepcopy` is involved there, and a group implementation could decide to cache the identity of the group, and implement both `one(G)` and `one(g)` to always return this canonical copy. What do we gain by effectively forbidding this?

But even for the first check: why should we care if it is "truly different"? I'd understand if we were talking about mutable data structures, there it is important to get a "properly distinct copy". But for e.g. a permutation... ?

**Comment by fingolfin** [comment]
@kalmarek BTW, thank you for reviewing these PRs and your comments -- I do appreciate them, and I am willing to have my mind changed, I just want to first understand thoroughly  :-)

**Comment by kalmarek** [comment]
@fingolfin thanks for taking the time;
yes, a better test would be to check if `deepcopy(g) !== g`. And I believe that this

> The purpose is to make sure that if a group element is deepcopiable then deepcopy_internal (which the user must implement) works as intended = a deepcopy of a deepcopiable element is truly distinct.

has real value of catching user errors. In the API we ask user to deepcopy parts of GroupElement structure, while preserving parent. It seems appealing to me to add some sanity checks on the correctness of user implemented functions.
Especially if mutating functions are applied to the element failing this check may result in bugs which are very hard to trace.

btw. both `AA.Perm`s and `PermutationGroups.Permutation` are mutable.

**Comment by fingolfin** [comment]
I am not sure I understand your last comment correctly, but let's take it one step at a time, and put a pin into the `deepcopy` check, and let's focus on this one:

So far you have not at all addressed what I wrote about the `one(G) !== one(g)` check. What is the rationale for having it in there, given that it is not related to `deepcopy` in any obvious way?

**Comment by kalmarek** [comment]
As I have written, the intention was to check that we obtain truly different elements, if it is possible.
Returning identically same (e.g. cached) identity for mutable elts when `similar(g)` defaults to `one(g)` is asking for trouble. You may ask for `one(G)` and be fine, but the next time you get something different, since the singleton/cached identity got just mutated in the `similar(g)` call in `mul!` two lines above.

**Comment by fingolfin** [comment]
If that is the intention, then the condition should be to check ismutable(type), not deepcopy

**Comment by kalmarek** [comment]
> If that is the intention, then the condition should be to check ismutable(type), not isdeepcopyable.

First of all, we're not testing that an element can be mutated, but whether we can create a distinct, unrelated copy of it.

Secondly, `ismutable(type)` would be incorrect, as (immutable) `structs` can be mutated. If you add a full constructor to your `B`, then `ismutabletype(B)==false`, but of course `b=B(0,[1]); deepcopy(b) !== b`.

Finally, `_is_deepcopiable` is an undocumented, hidden function used only for guarding a few (2?) tests; It was introduced precisely so that GAPGroupElements can override it and pass the test suite. I asked a few times, but let me try again: what happens when you define `GroupsCore._is_deepcopiable(::GAPGroupElemen) = false` ? Does it work as intended?

**Comment by kalmarek** [comment]
I guess this is no longer relevant (and has been mostly addressed by #52. 
When needed feel free to reopen.
